### ISSUE TITLE
MiniMax-H3 Support Roadmap

### ISSUE THREAD

**Issue by kohya-ss** [kohya-ss/musubi-tuner#1029]
This issue tracks completed, planned, and under-consideration work for MiniMax-H3 support. It is referenced from the `dev` → `main` draft PR (#1030), which shows what is queued for the next release.

**If you are considering contributing a feature in this area, please check this list and leave a comment here (or open an issue) before starting substantial work** — some items are already planned or in progress on our side, and we would like to avoid duplicated effort.

## Done (merged into `dev`)

- [x] Initial MiniMax-H3 support: T2VA / FL2VA / Ref2VA training and inference (#1018 — thanks @sdbds)
- [x] ConvRot INT8 quantized base weights: on-the-fly quantization of BF16 checkpoints, and automatic detection of ComfyUI pre-quantized ConvRot INT8 artifacts (transformer and text encoder) (#1024, #1025)
- [x] State dict loading unification and trainer cleanup (#1026, #1028)
- [x] **NVFP4 support** — W4A4 compute using `torch.nn.functional.scaled_mm` (PyTorch 2.10+, hardware-accelerated on Blackwell), with a dequantize fallback for older GPUs and PyTorch versions. An experimental implementation exists on a separate branch and will be adapted to MiniMax-H3. (#1031)
- [x] **Loading pre-quantized ComfyUI artifacts** — NVFP4/AWQ Qwen3-VL text encoder and tensorwise INT8 checkpoints (see #1027 for background — thanks @spawner1145 for demonstrating the demand and the format details). (#1031)
- [x] **Text encoder streaming** — reduce peak memory while loading/encoding with the Qwen3-VL text encoder. (#1032)
- [x] Recover bounded audio pts jitter in `decode_audio`. (#1033)
- [x] Opt-opt audio for reference video (e.g. motion reference). (#1034)
- [x] Refactor timestep sampling to get base timesteps. (#1035)
- [x] Refactor RNG generation and augmentation in training. (#1036)
- [x] Fix tokenizer setting to use the official MiniMax H3 tokenizer. (#1040)
- [x] **Support guidance loss** (contrastive loss). (#1045)
- [x] AdaLN-pruned checkpoint self-conversion from full BF16 checkpoints (pruned BF16 files doesn't have time_embedder) (#1046)
- [x] FL2VA/REF2VA teacher matching training (#1047)
- [x] Improve VAE image decoding and support One-frame (image) generation (#1054, #1055, #1056)
- [x] Image only training for t2va (#1057)
- [x] Image only training for fl2va (Edit LoRA) (related #1044, commented on [here](#issuecomment-5252550980)) (#1058)

## Planned / in progress

These items are already being worked on. Please discuss before submitting PRs for them.

- [ ] Follow-up quality review of the ConvRot INT8 implementation (numerics validation).
- [x] Image only training with teacher maching loss.

## Under consideration

No committed timeline; design and feasibility are still being evaluated.

- [ ] Support I2VA and L2VA training: If we have training videos, FL2VA training is always possible, so this seems like it can be a low priority.

## Notes

- Current limitations are documented in [`docs/minimax_h3.md`](../blob/dev/docs/minimax_h3.md#limitations).
- This list will be updated as work progresses. Feel free to comment with requests or questions.

**Comment by sdbds** [comment]
https://github.com/sdbds/musubi-tuner/pull/2
After reviewing it, I think the image training part can be directly ported over, and the quantization part can be removed.

**Comment by lukemcook** [comment]
Thanks for your work on this. Just wondering about the modes, is FL2VA intended to be an I2V training mode? The Comfy templates for I2V use only the first frame part and leave the last frame disconnected, so it seems like FL2VA is indeed also a regular I2V mode but wondering if you have any insight. Thanks again!

**Comment by kohya-ss** [comment]
FL2VA model seems to be intended as T2VA, I2VA and L2VA.

https://huggingface.co/MiniMaxAI/MiniMax-H3/blob/main/docs/VIDEO_PROMPT_WRITING_GUIDE_base_en.md

**Comment by lukemcook** [comment]
> FL2VA model seems to be intended as T2VA, I2VA and L2VA.
> 
> https://huggingface.co/MiniMaxAI/MiniMax-H3/blob/main/docs/VIDEO_PROMPT_WRITING_GUIDE_base_en.md

Ah yes but I mean specifically --task FL2VA vs. --task T2VA! I'm not sure if this is like LTX where the T2V mode is the same as I2V with a conditioning start image or if FL2VA mode is intended for first frame and last frame images instead of only a first frame (ordinary I2V) starting point

**Comment by kohya-ss** [comment]
> > FL2VA model seems to be intended as T2VA, I2VA and L2VA.
> > https://huggingface.co/MiniMaxAI/MiniMax-H3/blob/main/docs/VIDEO_PROMPT_WRITING_GUIDE_base_en.md
> 
> Ah yes but I mean specifically --task FL2VA vs. --task T2VA! I'm not sure if this is like LTX where the T2V mode is the same as I2V with a conditioning start image or if FL2VA mode is intended for first frame and last frame images instead of only a first frame (ordinary I2V) starting point

FL2VA model is versatile: it functions as T2VA if no image is specified, as I2VA if only the first frame is specified, and as L2VA if only the last frame is specified.

Since this PR currently supports only T2VA and FL2VA, I would like to add support for I2VA and L2VA as well.

**Comment by lukemcook** [comment]
> > > FL2VA model seems to be intended as T2VA, I2VA and L2VA.
> > > https://huggingface.co/MiniMaxAI/MiniMax-H3/blob/main/docs/VIDEO_PROMPT_WRITING_GUIDE_base_en.md
> > 
> > 
> > Ah yes but I mean specifically --task FL2VA vs. --task T2VA! I'm not sure if this is like LTX where the T2V mode is the same as I2V with a conditioning start image or if FL2VA mode is intended for first frame and last frame images instead of only a first frame (ordinary I2V) starting point
> 
> FL2VA model is versatile: it functions as T2VA if no image is specified, as I2VA if only the first frame is specified, and as L2VA if only the last frame is specified.
> 
> Since this PR currently supports only T2VA and FL2VA, I would like to add support for I2VA and L2VA as well.

Thanks, then it seems --task FL2VA is probably more ideal for an I2V focused LORA

**Comment by kohya-ss** [comment]
## MiniMax-H3 one-frame (image) generation and LoRA training — design summary

This is the plan for H3 still-image support (generation and LoRA training). It is based on
experiments we ran on the released model before reviewing any related PRs; the full design doc
will land with the implementation PRs. Related: PR #1044 (closed — see the discussion there;
three of its ideas will be adopted separately with credit).

### Experimental grounding

- **The video VAE's single-latent-token decode path is broken** (~15 dB PSNR, severe
  hue/scanline artifacts): the ViT3D decoder needs at least 2 tokens of temporal context.
  Duplicating the single token to a pseudo-T=2 sequence, running the normal temporal decode,
  and keeping frame 0 recovers to within ~1–2 dB of a true 2-token decode and is visually
  clean. A single-image encode is exactly the first latent token of a longer static clip
  (causal encoder), so the encode side needs no changes.
- **A true `T_lat=1` video target works at inference**: the transformer denoises a 1-token
  target it never saw in training, at quality comparable to the shortest on-grid clip
  (`frame_count=5`, 2 tokens). Plain t2va (no visual conditions) produces clean stills at both
  lengths.
- Both of the above were also found independently in the excellent technical report on
  `latent_t=1` generation by 月須和・那々 (nana_tsukisuwa on X) — including the same
  duplicate-then-decode fix, refs-only conditioning, and a demonstration that LoRA training at
  `T_lat=1` adapts correctly. Our measurements reproduce theirs.
- **Timestamps are a discrete anchor-snapping signal, not a continuous one**: placing a 1-token
  target at exactly a condition's RoPE timestamp reproduces that condition almost pixel-exactly
  (RMS ≈ 0.012), while intermediate positions do not interpolate on the base model. The
  positional pathway works; continuous behavior is something a LoRA can be taught. This also
  means training data must NOT put condition images at the target's own timestamps — that
  fights the model's strongest prior.

### Design: a unified "one-frame mode"

A `T_lat=1` target plus K single-frame visual conditions with explicit timestamps:

| K | task | use case |
|---|------|----------|
| 0 | t2va | plain image LoRA / image generation |
| 1 | fl2va (first only) | image-editing LoRA (relight, style transfer, camera move, ...) |
| 2 | fl2va (first+last) | inbetweening LoRA, two-anchor edits |

Key decisions:

- **Time unit = 24 fps pixel-frame index** (`control_index` / `target_index`, integers), mapped
  exactly onto H3's clock (`5/3` RoPE units per frame). Same shape as Wan's one-frame training;
  conditions keep their natural timestamps and the target sits elsewhere.
- **Explicit mode, not a global relaxation**: `--one_frame` on cache/train, `--frame_count 1`
  on generation. The `17n+5` / `5n+2` validators stay strict outside it. Layout timestamps
  become part of the packed-layout value (optional overrides field), so the rotary cache stays
  correct by construction.
- **Caches stay time-agnostic**: latents carry no time; indices live in cache metadata and the
  layout is built at runtime. Dataset-fixed indices first; per-item JSONL indices later
  (needed for inbetween training, where every sample has its own α); runtime index
  randomization stays possible without recaching.
- **Dataset side reuses existing machinery**: `image_directory` / `image_jsonl_file` (+
  `control_directory` and the `fp_1f_*` index keys, as in Wan/FramePack one-frame). Audio rows
  use the existing unsupervised silence placeholder (2 latent frames for a 1-frame target).
- **VAE fix ships first**: the single-token decode branch becomes duplicate-to-2 + frame 0.
  It has no production consumers today, so this is a pure improvement.
- Image outputs are written as PNG (audio decode skipped); `frame_count=5` generation stays
  available unchanged.

### Phasing

- **Phase 0**: VAE single-token decode fix (standalone).
- **Phase 1**: K=0 end-to-end — image datasets, one-frame cache/train/sample/generate, docs.
  `batch_size=1` + gradient accumulation (multi-sample sequence packing with a block-diagonal
  attention mask is a separate future work item; it matters most for image training, where
  per-step fixed costs like block-swap traffic dominate).
- **Phase 2**: control images with indices (editing / inbetweening), per-item JSONL indices.
- **Phase 3**: ref2va one-frame (untimed references, image target).

Feedback welcome, especially from anyone who has trained image or inbetween LoRAs on other
video architectures.

**Comment by sdbds** [comment]
I almost forgot—I think we could try inferring the use of the sol-attn proxy for sage-attn?

**Comment by rockerBOO** [comment]
This is implying that convrot int8 and nvpf4 in-process quantization AND loading quantized models will both be working? This includes the in-process quantization for krea 2 from before but is nvfp4 supported for all models? Or is this just h3 for this PR?

**Comment by christopher5106** [comment]
We would like to contribute one-frame (image) training for `--task ref2va` — reference-driven image LoRAs, the mode docs/minimax_h3_1f.md currently lists as not implemented. Since one-frame ref2va *generation* already works, our understanding is the gap is mostly dataset/cache plumbing (image datasets carrying per-record references) plus lifting the training-side gates, with tests following the existing one-frame FL2VA patterns.

Before building it: are there design constraints you'd want respected? The two decisions we'd otherwise make ourselves: (1) dataset schema — extend `image_jsonl_file` records with the same ordered `references` list the video JSONL uses; (2) scope — image references only in the first pass (video/audio refs for a one-frame target left gated), guidance loss supported (per #1075), teacher-matching left out. Happy to adjust to whatever fits your plans for this list.

**Comment by YPatrickW** [comment]
Hello， We would like to fine-tune the Ref2VA model for video editing tasks, but the generated results appear to simply copy the original video, any suggestions?

**Comment by kohya-ss** [comment]
I plan to make changes tomorrow to the inference script `minimax_h3_generate_video.py
` arguments (such as `--output` etc.) to accommodate other architectures. Please note that this will be a breaking change.

**Comment by kohya-ss** [comment]
CLI is updated in #1107.

**Comment by christopher5106** [comment]
Question about the provenance of the 5 s lower bound on target duration, since it turns out to be the deciding constraint for Ref2VA with video references.

`minimax_h3_cache_latents.py` enforces:

```python
duration = Fraction(frame_count, TARGET_FPS)
if not allow_experimental_duration and not (Fraction(5, 1) <= duration <= Fraction(15, 1)):
    raise ValueError(f"MiniMax-H3 target duration {float(duration):.3f}s is outside the released 5-15s range; ...")
```

MiniMax's model card states **"Output duration: 4–15 seconds"**. Because frame counts must be 17n+5, exactly one step sits between the two readings:

| frames | duration | vs 5 s bound | vs 4 s bound |
| --- | --- | --- | --- |
| 90 (n=5) | 3.750 s | rejected | rejected |
| **107 (n=6)** | **4.458 s** | rejected | allowed |
| 124 (n=7) | 5.167 s | allowed | allowed |

So: is the 5 s floor taken from a released source that specifies 5 (a training config or inference constraint I have not found), or is it a deliberately conservative rounding of the documented 4–15 s output range? Put differently — would 107 frames be genuinely off-distribution, or legitimate-but-unverified?

**Why it matters concretely.** A Ref2VA target with a *video* reference is expensive, because the reference is capped to the target's duration and therefore costs a second sequence rather than a still. At 1792x768 with one video reference per entry:

- target alone: 49,728 rows
- target + video reference: 85,470 rows (1.72x)

That OOMs a 180 GB B200 during the first training step — 174.31 GiB of 178.35 in use, short by 5.07 GiB, in the LoRA MLP forward. The same dataset with **image** references (one frame instead of the target's full span) trains fine at 124 frames, so it is specifically the video-reference case.

At 107 frames the same configuration would be ~74k rows, which should fit comfortably. I tried 90 first and hit the duration guard, which is what prompted this question — 124 really is the smallest valid step under the 5 s rule.

Not asking for a change; `--allow_experimental_duration` already exists. I would just rather know whether using it here means "slightly outside what kohya chose to bless" or "genuinely outside what the model was trained on", before training anything at 107 and drawing conclusions from the result.

Happy to contribute the measurements if useful. Context: reforge is training H3 video LoRAs with one dataset block per aspect ratio, pinned to a 768 short side.
